master: nightly linux-race failed

The nightly linux-race test run failed on d86f6dab212c6fa9fb3c5f029554e49b536fc23a (go1.24). Please review the run [here](https://github.com/cockroachdb/pebble/actions/runs/17513130932).

```
--- FAIL: TestMarkedForCompaction (0.03s)
    logger.go:15: Found 0 WALs
    logger.go:15: Found 1 WALs
    logger.go:15:   - 000002: {(,000)}
    logger.go:15: [JOB 1] WAL 000002: {(,000)} stopped reading at offset: (000002.log: 0); replayed 0 keys in 0 batches
    compaction_test.go:2798: 
          testdata/marked_for_compaction:42:
          maybe-compact
        output didn't match expected:
          @@ -1,7 +1,6 @@
          -[JOB 100] compacted(rewrite) L1 [000006] (665B) Score=0.00 + L1 [] (0B) Score=0.00 -> L1 [000012] (665B), in 1.0s (2.0s total), output rate 665B/s
           [JOB 100] compacted(rewrite) L0 [000007] (654B) Score=0.00 + L0 [] (0B) Score=0.00 -> L0 [000013] (654B), in 1.0s (2.0s total), output rate 654B/s
           L0.0:
             000013:[c#11,SET-c#11,SET] seqnums:[11-11] points:[c#11,SET-c#11,SET] size:654
           L1:
          -  000012:[c#0,SET-d#0,SET] seqnums:[0-0] points:[c#0,SET-d#0,SET] size:665
          +  000008:[c#0,SET-d#0,SET] seqnums:[0-0] points:[c#0,SET-d#0,SET] size:665
```

--- a/compaction_test.go
+++ b/compaction_test.go
@@ -2774,21 +2774,25 @@ func TestMarkedForCompaction(t *testing.T) {
 	var d *DB
 
 	var buf bytes.Buffer
+	eventListener := EventListener{
+		CompactionEnd: func(info CompactionInfo) {
+			// Fix the job ID and durations for determinism.
+			info.JobID = 100
+			info.Duration = time.Second
+			info.TotalDuration = 2 * time.Second
+			fmt.Fprintln(&buf, info)
+		},
+	}
+	if testing.Verbose() {
+		eventListener = TeeEventListener(eventListener, MakeLoggingEventListener(base.DefaultLogger))
+	}
 	opts := &Options{
 		FS:                          mem,
 		DebugCheck:                  DebugCheckLevels,
 		DisableAutomaticCompactions: true,
 		FormatMajorVersion:          internalFormatNewest,
-		EventListener: &EventListener{
-			CompactionEnd: func(info CompactionInfo) {
-				// Fix the job ID and durations for determinism.
-				info.JobID = 100
-				info.Duration = time.Second
-				info.TotalDuration = 2 * time.Second
-				fmt.Fprintln(&buf, info)
-			},
-		},
-		Logger: testutils.Logger{T: t},
+		EventListener:               &eventListener,
+		Logger:                      testutils.Logger{T: t},
 	}
 	opts.Experimental.CompactionScheduler = func() CompactionScheduler {
 		return NewConcurrencyLimitSchedulerWithNoPeriodicGrantingForTest()
@@ -2869,6 +2873,9 @@ func TestMarkedForCompaction(t *testing.T) {
 			d.mu.Lock()
 			defer d.mu.Unlock()
 			d.opts.DisableAutomaticCompactions = false
+			defer func() {
+				d.opts.DisableAutomaticCompactions = true
+			}()
 			for {
 				d.maybeScheduleCompaction()
 				if d.mu.compact.compactingCount == 0 {
@@ -2881,7 +2888,6 @@ func TestMarkedForCompaction(t *testing.T) {
 
 			fmt.Fprintln(&buf, d.mu.versions.currentVersion().DebugString())
 			s := strings.TrimSpace(buf.String())
-			opts.DisableAutomaticCompactions = true
 			return s
 
 		default:

--- a/options.go
+++ b/options.go
@@ -987,6 +987,9 @@ type Options struct {
 	// DisableAutomaticCompactions dictates whether automatic compactions are
 	// scheduled or not. The default is false (enabled). This option is only used
 	// externally when running a manual compaction, and internally for tests.
+	//
+	// Note that this field is only consulted while holding DB.mu, so it it safe
+	// for a test to modify it while holding DB.mu.
 	DisableAutomaticCompactions bool
 
 	// DisableConsistencyCheck disables the consistency check that is performed on

